experiment/simulation/js/fa.js

```diff
@@ -0,0 +1,333 @@
+        this.sum = (getOutputFA(this.Cin[0], this.Cin[1]) && !aXorb) || (!(getOutputFA(this.Cin[0], this.Cin[1])) && aXorb);
+    }
+
+    // Sets the output enpoint of the full adder as connected
+    setConnected(val, pos) {
+        console.log(val, pos);
+        if (pos === "Carry") {
+            this.CoutIsConnected = val;
+        }
+        else if (pos === "Sum") {
+            this.sumIsConnected = val;
+        }
+    }
+
+}
+
+// Add a full adder to the circuit board
+function addFA() {
+    let fA = new FullAdder();
+    fA.registerComponent("working-area");
+}
+
+window.AddFA = addFA;
+
+// Used to extract output from a given gate, if pos isnt empty the gate is a full adder with the position specified
+function getOutputFA(gate, pos) {
+    if (pos === "Carry") {
+        return gate.Cout;
+    }
+    else if (pos === "Sum") {
+        return gate.sum;
+    }
+    // But if the gate is not an FA, but an input bit, then return the value of the input
+    else {
+        return gate.output;
+    }
+}
+
+
+
+// Recursive function that evaluates the output of the full adder
+function getResultFA(fa) {
+// check if fa type is Gate object
+    if (fa.constructor.name === "Gate") {
+        return;
+    }
+
+    if (fa.Cout != null && fa.sum != null) {
+        return;
+    }
+
+    if (getOutputFA(fa.a0[0], fa.a0[1]) == null) {
+        getResultFA(fa.a0[0]);
+    }
+    if (getOutputFA(fa.b0[0], fa.b0[1]) == null) {
+        getResultFA(fa.b0[0]);
+    }
+    if (getOutputFA(fa.Cin[0], fa.Cin[1]) == null) {
+        getResultFA(fa.Cin[0]);
+    }
+
+    fa.generateOutput();
+
+    return;
+}
+
+
+// Checks if the connections are correct
+function checkConnectionsFA() {
+    let flag = 0;
+    for (let faID in fullAdder) {
+        const gate = fullAdder[faID];
+        // For Full Adder objects
+        // Check if all the outputs are connected
+        if (gate.CoutIsConnected === false) {
+            flag = 1;
+            break;
+        }
+        if (gate.sumIsConnected === false) {
+            flag = 1;
+            break;
+        }
+
+        // Check if all the inputs are connected
+        if (gate.a0 == null || gate.a0.length === 0) {
+            flag = 1;
+            break;
+        }
+        if (gate.b0 == null || gate.b0.length === 0) {
+            flag = 1;
+            break;
+        }
+        if (gate.Cin == null || gate.Cin.length === 0) {
+            flag = 1;
+            break;
+        }
+    }
+    for (let gateId in gates) {
+        const gate = gates[gateId];
+        if (gate.isInput === true) {
+            if (gate.isConnected === false) {
+                flag = 1;
+                break;
+            }
+        }
+        if (gate.isOutput === true) {
+            if (gate.inputs.length === 0) {
+                flag = 1;
+                break;
+            }
+        }
+    }
+
+
+
+    if (flag === 0) {
+        return true;
+    }
+    else {
+        alert("Connections are not correct");

```

Comment:
Are we planning to use alert for warnings/errors?

Reply:
Yes, we had agreed to that in the weekly meet if I recall correctly.